EIPS/eip-3607.md

```diff
@@ -0,0 +1,94 @@
+
+This EIP is to specify this behaviour to always forbid such transactions. This fixes most realistic or serious attacks due to address collisions.
+
+
+## Specification
+
+Any transaction where `tx.sender` has a `CODESIZE != 0` MUST be rejected as invalid.
+
+## Rationale
+
+We note that it was always the expected that a contract account's behaviour is constrainted by the code in that contract -- which means that the account's funds should not suddenly be spendable by some private key. It was just implicitly assumed in the past that a 160 bit address length is enough to provide collision resistance, and thus that this case could never occur. In that sense, this EIP should be seen as a clarification of protocol behaviour in a previously undefined case rather than an explicit upgrade of consensus rules.
+
+This does not exclude all possible attack vectors, only the most serious one. Further possible attack vectors via address collisions between contracts and EOAs are:
+1. An attacker can convince a user to send funds to an account before it is deployed. Some applications require this behaviour (e.g. state channels).
+2. A chain reorg can happen after a contract is deployed. If the reorg removes the contract deployment transaction the funds can still be accessed using the private key.
+3. A contract can self desctruct, with the stated intention that ERC20s (or other tokens) in the contract would be burned. However, they can now be accessed by a key for that address.
+
+All these scenarios are much harder to exploit for an attacker, and likely have much lower yield making the attacks unlikely to be economically viable.
+
+## Backwards Compatibility
+
+It is unlikely that an attack like this has already occurred on the Ethereum mainnet, or we would very likely have heard of it. It is inconceivable that someone would use this as a "feature" to make a contract an EOA at the same time, when they could simply do this by adding some methods to the contract instead of spending billions on building hardware to find hash collisions.
+
+Private networks may have deployed contracts which also work as EOAs at genesis and should check that this upgrade does not impact their workflows.
+
+## Test Cases
+
+Given a genesis allocation of  
+```
+Address: 0x71562b71999873DB5b286dF957af199Ec94617F7
+Balance: 1000000000000000000 // 1 ether
+Nonce:   0,
+Code:    0xB0B0FACE",
+```
+Every transaction send by the private key corresponding to `0x715656...` (
+`b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291`) fails.
+
+A set of test cases can be found [here](https://github.com/ethereum/tests/pull/879)
+
+## Reference Implementation
+
+The following check must be added to the state transition checks after checking that the nonce of the sender is correct.
+```go
+// Make sure the sender is an EOA
+if codesize := st.state.GetCodeSize(st.msg.From()); codesize != 0 {
+	return ErrSenderNoEOA
+}
+```
```

Comment:
Consider writing this in pseudocode or a more well known/easier to read language than Go.  Also, consider extracting the assignment out of the conditional for a bit more clarity/readability.

Reply:
@MariusVanDerWijden